### PR TITLE
proposal: refactor quickstart and docs

### PR DESCRIPTION
**Pull request by slack**
Skeleton refactor to test new provider-specific flows.

One of the struggles I've had with the quickstart is that it has been
caught between among a task-oriented doc, platform reference and provider
agnostic. The end-result is something that isn't compelling and ends up
being particularly confusing for new visitors to the project.

This (now squashed) commit represents a deep refactor of the quickstart
documentation.

Orient the quickstart around a three step, goal-oriented process:

1. Tools
2. Infra + Workflow Install
3. Deploy an example app

This also pulls in specific configurations of Kubernetes for evaluation and
dev/test (kube-up.sh and GKE). We want tire-kickers to get a win and we
need to offer as much assistance through the process as possible.

Pre-install helm and deis CLIs up front.

Couching the install in provider-specific workflows gives the documentation
much needed context. For example, nuances about load balancer
configurations (or lack there-of) can be dealt with in-context. This will
simplify the language as we can back away from routing folks to more
detailed explanations or landing a barrage of caveats.

I also slimmed down the "deploy your first app" to the absolute basics.
Using the `deis/example-go` app lets folks get a win without having to deal
with cloning repos, etc. That stuff can go into reference documentation for
the three build types.

Click through the GKE flow, that should be complete.

Thoughts would be great! Then we can do sections for Vagrant and AWS.

### CODE REVIEW

**Comment by slack** [comment]
Finished up all three quickstart paths, would like to start getting eyes on this one.

File: src/index.md [hunk 1]
```diff
@@ -1,13 +1,25 @@
 # Welcome to Deis Workflow
 
-Deis (pronounced DAY-iss) Workflow is an open source Platform as a Service (PaaS) that adds a developer-friendly layer to any Kubernetes cluster, making it easy to deploy and manage applications on your own servers.
+Deis (pronounced DAY-iss) Workflow is an open source Platform as a Service (PaaS) that adds a developer-friendly layer
+to any Kubernetes cluster, making it easy to deploy and manage applications.
 
-Workflow includes capabilities for building and deploying from source via `git push`, configuring your applications, creating and rolling back releases, managing application domain names and SSL certificates, providing edge routing, aggregating application logs, and controlling access to application resources.  All of this is exposed through a simple REST API and command line interface.
+Deis Workflow includes capabilities for building and deploying from source via `git push`, configuring your applications,
```
Comment (mboersma) [inline src/index.md]: This sentence repeats "application" four times. Maybe omit a couple?
> Deis Workflow includes capabilities for building and deploying from source via `git push`, configuring your applications, creating and rolling back releases, managing domain names and SSL certificates, providing edge routing, aggregating logs, and controlling access to application resources.

File: src/index.md [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+creating and rolling back releases, managing application domain names and SSL certificates, providing edge routing,
+aggregating application logs, and controlling access to application resources. All of this is exposed through a simple
+REST API and command line interface.
 
-For more detailed information about Workflow, view the [Concepts][concepts] and [Architecture][arch] sections.
+## Getting Started
+
+Take a deep dive into Deis Workflow in our [Concepts][concepts], [Architecture][arch], and [Components][components]
```
Comment (mboersma) [inline src/index.md]: While it can sometimes feel more conversational or friendly, using first-person pronouns like "we" and "our" also creates a semantic distinction I don't think is appropriate for an open source project. I prefer:
> Take a deep dive into Deis Workflow in **the** [Concepts][concepts], [Architecture][arch], ...

It's a finer point, however. And it may be pervasive. If others disagree feel free to ignore me.

File: src/index.md [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+sections.
 
 To get started with Workflow, follow our [Quick Start][quickstart] guide.
 
+## Service and Support
+
+If you are interested in commercial service and support for Deis Workflow, check out our [services offerings on deis.com](https://deis.com/services).
```
Comment (mboersma) [inline src/index.md]: As above, I think "the" is better than "our" here. (And ideally this could be rephrased not to use "you." I don't want to make things sterile, just professional.)

File: src/installing-workflow/index.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+Elasticsearch is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/elasticsearch-logging
+Heapster is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/heapster
+Kibana is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/kibana-logging
+KubeDNS is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/kube-dns
+kubernetes-dashboard is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard
+Grafana is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/monitoring-grafana
+InfluxDB is running at https://52.9.206.49/api/v1/proxy/namespaces/kube-system/services/monitoring-influxdb
+```
+
+If you see a list of targets like the one above, `helmc` can communicate with the Kubernetes master. Also make sure that
```
Comment (mboersma) [inline src/installing-workflow/index.md]: Maybe rephrase the second sentence:
> Double-check that the master matches the expected cluster.

File: src/installing-workflow/index.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+$ helmc --version
+helmc version 0.7.0+20a7ed7
+```
+
+Ensure the `kubectl` client is installed and can connect to your Kubernetes cluster. `helmc` uses `kubectl` to interact
+with your Kubernetes cluster.
+
+You can test that helmc and kubectl are working properly by running:
```
Comment (mboersma) [inline src/installing-workflow/index.md]: Should these commands be backticked for consistency?

> You can test that `helmc` and `kubectl` are working properly...

File: src/installing-workflow/index.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+You'll need to wait for the pods that it launched to be ready. Monitor their status
+by running:
+
+```
+$ kubectl get pods --namespace=deis
+```
+
+If you would like `kubectl` to automatically update as the pod states change, run (type Ctrl-C to stop the watch):
+```
+$ kubectl get pods --namespace=deis -w
```
Comment (mboersma) [inline src/installing-workflow/index.md]: Minor point: if you always put the `--namespace=` flag immediately after `kubectl` it makes it easier to alias:
```console
$ kubectl --namespace=deis get pods --watch
$ alias kd='kubectl --namespace=deis'
$ kd get po -w
```

File: src/index.md (hunk 2 above)
Comment (bacongobbler) [inline src/index.md]: I've always agreed with this sentiment. +1 for communicating docs in third person.

File: src/installing-workflow/index.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+Depending on the order in which the Workflow components start, you may see a few components restart. This is common during the installation process, if a component's dependencies are not yet available the component will exit and Kubernetes will automatically restart the containers.
```
Comment (mboersma) [inline src/installing-workflow/index.md]: Maybe:
> Depending on the order in which Workflow components initialize, some pods may restart. This is common during installation: if a component's dependencies are not yet available, the component will exit and Kubernetes will automatically restart it.

File: src/installing-workflow/system-requirements.md [hunk 8]
```diff
@@ -1,34 +1,45 @@
-# Resource Requirements
+# Requirements
 
-When deploying Deis, it's important to provision machines with adequate resources. Deis is a highly-available distributed system, which means that Deis components and your deployed applications will move around the cluster onto healthy hosts as hosts leave the cluster for various reasons (failures, reboots, autoscalers, etc.). Because of this, you should have ample spare resources on any machine in your cluster to withstand the additional load of running services for failed machines.
+If you are planning to run Deis Workflow on your own Kubernetes cluster there are a few requirements that you should be
```
Comment (mboersma) [inline src/installing-workflow/system-requirements.md]: Maybe:
> To run Deis Workflow on a Kubernetes cluster, there are a few requirements to keep in mind.

File: src/installing-workflow/system-requirements.md [hunk 9]
```diff
@@ -1,34 +1,45 @@
...
+## Resource Requirements
 
-Deis components consume approximately 2 - 2.5GB of memory across the cluster, and approximately 30GB of hard disk space. Because each machine should be able to absorb additional load should a machine fail, each machine must have:
+When deploying Deis Workflow, it's important to provision machines with adequate resources. Deis is a highly-available
+distributed system, which means that Deis components and your deployed applications will move around the cluster onto
+healthy hosts as hosts leave the cluster for various reasons (failures, reboots, autoscalers, etc.). Because of this,
+you should have ample spare resources on any machine in your cluster to withstand the additional load of running
+services for failed machines.
+
+Deis Workflow components consume approximately 2 - 2.5GB of memory across the cluster, and approximately 30GB of hard
```
Comment (mboersma) [inline src/installing-workflow/system-requirements.md]: Some repetition here. Maybe:
> Deis Workflow components use about 2.5GB of memory across the cluster, and require approximately 30GB of hard disk space. Because it may need to handle additional load if another one fails, the minimum requirements for each machine are:

File: src/installing-workflow/system-requirements.md [hunk 10]
```diff
@@ -1,34 +1,45 @@
...
+Docker daemons configured to use an **insecure registry**. The configured subnet should encompass any private networks
+used by your worker nodes, including overlay networks.
+
+Depending on your Kubernetes and Docker configuration, setting `EXTRA_DOCKER_OPTS="--insecure-registry=10.0.0.0/8"` may
+be sufficient.
 
 ## SELinux + OverlayFS
 
-If you are using Docker with OverlayFS you must disable SELinux by adding `--selinux-enabled=false` to `EXTRA_DOCKER_OPTS`. For more background information see:
+If you are using Docker with OverlayFS you must disable SELinux by adding `--selinux-enabled=false` to
```
Comment (mboersma) [inline src/installing-workflow/system-requirements.md]: I think both these sentences should have commas:
> If you are using Docker with OverlayFS, you must disable..
> For more background information, see:

File: src/managing-workflow/component-configuration.md [hunk 11]
```diff
@@ -19,7 +19,6 @@ env:
     value: "admin_only"
 ```
 
-
```
Comment (mboersma) [inline src/managing-workflow/component-configuration.md]: We had a convention of two blank lines before a Markdown header to make it easier to find, just FYI. It's not a big deal either way and we're probably not being consistent, so :shrug:.

File: src/managing-workflow/platform-monitoring.md [hunk 12]
```diff
@@ -34,18 +35,34 @@ With the release of Workflow Beta4 we now include a monitoring stack for introsp
 ```
 
 ### Grafana
-We expose Grafana through the router using [service annotations](https://github.com/deis/router#how-it-works). This allows users to access the Grafana UI by accessing `grafana.mydomain.com`. While  we provide a default username/password of `admin/admin` this can be overridden at any time by setting the following environment variables in `$CHART_HOME/workspace/workflow-$WORKFLOW_RELEASE/manifests/deis-monitor-grafana-rc.yaml`: `GRAFANA_USER` and `GRAFANA_PASSWD`.
 
-It will preload several dashboards that we've created to help operators get started with monitoring their Kubernetes and Workflow installations. Each dashboard is meant to be a starting place for the operator and is not representative of all the dashboards needed to monitor a production installation.
+We expose Grafana through the router using [service annotations](https://github.com/deis/router#how-it-works). This
```
Comment (mboersma) [inline src/managing-workflow/platform-monitoring.md]: Maybe:
> **Workflow** exposes Grafana through the router...

File: src/managing-workflow/platform-monitoring.md [hunk 13]
```diff
@@ -34,18 +35,34 @@ With the release of Workflow Beta4 we now include a monitoring stack for introsp
...
+allows users to access the Grafana UI by accessing `grafana.mydomain.com`. While  we provide a default username/password
```
Comment (mboersma) [inline src/managing-workflow/platform-monitoring.md]: Repetition here. Maybe:
> This allows users to access the Grafana UI **at** `grafana.mydomain.com`.

And:
> The default username/password of `admin/admin` can be overridden at any time...

File: src/managing-workflow/platform-monitoring.md [hunk 14]
```diff
@@ -34,18 +35,34 @@ With the release of Workflow Beta4 we now include a monitoring stack for introsp
...
+of `admin/admin` this can be overridden at any time by setting the following environment variables in
+`$CHART_HOME/workspace/workflow-$WORKFLOW_RELEASE/manifests/deis-monitor-grafana-rc.yaml`: `GRAFANA_USER` and
+`GRAFANA_PASSWD`.
+
+It will preload several dashboards that we've created to help operators get started with monitoring their Kubernetes and
```
Comment (mboersma) [inline src/managing-workflow/platform-monitoring.md]: Maybe:
> Grafana will preload several dashboards to help operators get started with monitoring Kubernetes and Deis Workflow. These dashboards are meant as starting points and don't include every item that might be desirable to monitor in a production installation.

File: src/managing-workflow/platform-monitoring.md [hunk 15]
```diff
@@ -34,18 +35,34 @@ With the release of Workflow Beta4 we now include a monitoring stack for introsp
...
+Workflow installations. Each dashboard is meant to be a starting place for the operator and is not representative of all
+the dashboards needed to monitor a production installation.
 
-We are currently not writing the data to the host file system or to long term storage. Therefore, if the Grafana instance dies you will lose all custom and modified dashboards. It is recommended that you export your dashboards and store them in version control until a solution is implemented for long term storage.
+We are currently not writing the data to the host file system or to long term storage. Therefore, if the Grafana
```
Comment (mboersma) [inline src/managing-workflow/platform-monitoring.md]: Maybe:
> Workflow monitoring does not currently write data to the host filesystem or to long-term storage. If the Grafana instance fails, modified dashboards are lost. Until there is a solution to persist this, export dashboards and store them separately in version control.

**Comment by slack** [comment]
Updated and rebased.

**Comment by bacongobbler** [comment]
manually testing this as well now :)

**Comment by bacongobbler** [comment]
what's with the TODO.md?

File: src/index.md [hunk 16]
```diff
@@ -1,13 +1,25 @@
...
+routing, aggregating logs, and sharing applications with teams. All of this is exposed through a simple REST API and
+command line interface.
 
-For more detailed information about Workflow, view the [Concepts][concepts] and [Architecture][arch] sections.
+## Getting Started
+
+Take a deep dive into Deis Workflow in our [Concepts][concepts], [Architecture][arch], and [Components][components]
+sections.
 
 To get started with Workflow, follow our [Quick Start][quickstart] guide.
```
Comment (bacongobbler) [inline src/index.md]: I'd move this sentence higher. Makes more sense to talk about quick start documentation first, then go into a deeper dive.

File: src/quickstart/index.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
+# Quick Start
+
+Getting started with Deis Workflow in three easy steps.
```
Comment (bacongobbler) [inline src/quickstart/index.md]: I'd change this to "To get started with Deis Workflow, there are three steps to follow:"

File: src/quickstart/index.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+1. Install CLI tools for Helm Classic and Deis Workflow
+2. Boot a Kubernetes and install Deis Workflow
+3. Deploy your first application
+
+This guide will help you setup a cluster suitable for evaluation, development and testing. When you are ready for staging and production, view our [production checklist](../managing-workflow/production-deployments.md).
```
Comment (bacongobbler) [inline src/quickstart/index.md]: "setup" should be two words: "set up".

File: src/managing-workflow/production-deployments.md [hunk 19]
```diff
@@ -1,3 +1,7 @@
 # Production Deployments
 
-TODO (bacongobbler): rewrite for v2
+1. Configure off-cluster object storage
```
Comment (bacongobbler) [inline src/managing-workflow/production-deployments.md]: might want to add more detail here, or just integrate https://github.com/deis/workflow/pull/141

File: src/quickstart/index.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+## Step 1: Install CLI tools
+
+For the quickstart we will [install both Helm Classic and Deis Workflow CLI](install-cli-tools.md).
```
Comment (bacongobbler) [inline src/quickstart/index.md]: We should quickly add some more info here about what these clients do from a high level either now or in a followup. "quick start" doesn't have to be sparse on information. :)

File: src/quickstart/index.md [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+
+## Step 2: Boot Kubernetes and Install Deis Workflow
```
Comment (bacongobbler) [inline src/quickstart/index.md]: Should we separate this into two steps? "boot kubernetes" and "install deis workflow" seem like two separate steps, since one directs you to the kubernetes docs and one directs you to helm-classic.

File: src/quickstart/index.md [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+your laptop.
+
+Cloud-based options:
+
+* [Google Container Engine](provider/gke/boot.md): provides a managed Kubernetes environment, available with a few clicks.
+* [Amazon Web Services](provider/aws/boot.md): uses Kubernetes upstream `kube-up.sh` to boot a cluster on AWS EC2.
+
+If you would like to test on your local machine follow our guide for [Vagrant](provider/vagrant/boot.md).
+
+If you have already created a Kubernetes cluster, check out the [system requirements](../installing-workflow/system-requirements.md) and then proceed to [installing Deis Workflow on your own Kubernetes clsuter](../installing-workflow/index.md).
```
Comment (bacongobbler) [inline src/quickstart/index.md]: I'd change this to "and then proceed to install Deis workflow" rather than "installing"

File: mkdocs.yml [hunk 23]
```diff
@@ -4,49 +4,63 @@
...
+  - Home: index.md
+  - Quick Start:
+    - Overview: quickstart/index.md
+    - Install CLI Tools: quickstart/install-cli-tools.md
+    - AWS:
+      - Boot: quickstart/provider/aws/boot.md
+      - DNS: quickstart/provider/aws/dns.md
+      - Install Workflow: quickstart/provider/aws/install-aws.md
+    - GKE:
+      - Boot: quickstart/provider/gke/boot.md
```
Comment (bacongobbler) [inline mkdocs.yml]: We should change the title here. "Boot" seems like an odd choice of words as a title to install a GKE cluster (or AWS)

File: src/quickstart/index.md [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+There are many ways to boot and run Kubernetes. You may choose to get up and running in cloud environments or locally on
+your laptop.
+
+Cloud-based options:
+
+* [Google Container Engine](provider/gke/boot.md): provides a managed Kubernetes environment, available with a few clicks.
+* [Amazon Web Services](provider/aws/boot.md): uses Kubernetes upstream `kube-up.sh` to boot a cluster on AWS EC2.
+
+If you would like to test on your local machine follow our guide for [Vagrant](provider/vagrant/boot.md).
```
Comment (bacongobbler) [inline src/quickstart/index.md]: your local machine,* follow

**Comment by bacongobbler** [comment]
I also cannot click on the AWS, GKE or Vagrant selections on the right-hand-side dropdown list. This is likely due to a limitation with the number of levels the table of contents can handle, so I'd put the "Boot" sections as the top level index and make those links clickable.

File: mkdocs.yml [hunk 25]
```diff
@@ -4,49 +4,63 @@
...
-    - Registering a User: using-workflow/registering-a-user.md
-    - Deploying an Application: using-workflow/deploying-an-application.md
-    - Process Types and the Procfile: using-workflow/process-types-and-the-procfile.md
-    - Using Buildpacks: using-workflow/using-buildpacks.md
-    - Using Dockerfiles: using-workflow/using-dockerfiles.md
-    - Using Docker Images: using-workflow/using-docker-images.md
-    - Configuring an Application: using-workflow/configuring-an-application.md
-    - Managing an Application: using-workflow/managing-an-application.md
-    - Application SSL Certificates: using-workflow/application-ssl-certificates.md
+    - Command Line Interface: using-workflow/cli.md
```
Comment (bacongobbler) [inline mkdocs.yml]: this is a confusing title choice, since "users and registration" and "user ssh keys" both use the command line interface. Is there any other way we would use Workflow other than through the CLI?

File: mkdocs.yml [hunk 26]
```diff
@@ -4,49 +4,63 @@
...
+    - Users and Registration: using-workflow/users-and-registration.md
+    - User SSH Keys: using-workflow/user-ssh-keys.md
+  - Applications:
```
Comment (bacongobbler) [inline mkdocs.yml]: Perhaps this should all be under a reference section like [heroku](https://devcenter.heroku.com/categories/reference), or under a "User's Guide" section. Seems confusing to have a top level header for managing Applications, but none for other top level commands like managing Users (or eventually service brokers).

File: src/quickstart/index.md (hunk 21 above)
Comment (bacongobbler) [inline src/quickstart/index.md]: Ah, I remember why now. Each provider has provider-specific docs on getting it backed by S3 or GCS. Never mind!

File: mkdocs.yml (hunk 25 above)
Comment (slack) [inline mkdocs.yml]: The goal was to consolidate information about the CLI into one place. Installation, basic usage, proxy support and multiple profiles, etc. Less about bringing _all_ CLI commands into one doc. Make sense?

File: mkdocs.yml (hunk 26 above)
Comment (slack) [inline mkdocs.yml]: I ordered the stuff under the application section so that you could have a logical flow of detail...

Start with deploy overview (which still needs work), then the build types, then process management, configuration, lifecycle (which I don't like and should be split out to scaling), etc...

2016-05-17 at 16 04

Does that track for you?

File: src/managing-workflow/production-deployments.md (hunk 19 above)
Comment (slack) [inline src/managing-workflow/production-deployments.md]: Will integrate when either PR lands.

**Comment by slack** [comment]
Regarding the busted links, the whole CSS/TOC thing is a disaster and I'll tackle that separately.

File: mkdocs.yml (hunk 25 above)
Comment (bacongobbler) [inline mkdocs.yml]: makes sense. I guess this means we're going to need to break this up eventually if the user guide is specifally just about cli commands to "use workflow".

File: mkdocs.yml (hunk 26 above)
Comment (bacongobbler) [inline mkdocs.yml]: Perhaps it would make more sense to consolidate this all under "Using Workflow"? Managing your application lifecycle or deploying apps sounds like something a platform user would want to know about. Not sure if it's necessary to break it out into another section.

Comment (slack) [inline mkdocs.yml]: Well all of it belongs in "Using Workflow" when it comes down to it. I'm looking for a logical grouping so you don't have to go hunting in a 16 foot list of doc titles.

If we renamed "Using Workflow" to something like "Users" would that help?

Comment (bacongobbler) [inline mkdocs.yml]: If only we could have three levels of indirection so we could do "Using Workflow" -> "Applications" -> "Deploying Apps".

Yeah I think renaming Using Workflow would be ok. If we keep "Using Workflow" then I'm just going to want to re-org this back into that umbrella.

File: src/managing-workflow/component-configuration.md (hunk 11 above)
Comment (bacongobbler) [inline src/managing-workflow/component-configuration.md]: I think we need to find a markdown linter in the future if we want to enforce consistent styling. :)

File: src/managing-workflow/production-deployments.md (hunk 19 above)
Comment (bacongobbler) [inline src/managing-workflow/production-deployments.md]: could we remove these changes so rebasing #141 becomes easier?

File: themes/deis/base.html [hunk 27]
```diff
@@ -27,27 +27,7 @@
         <![endif]-->
     <div class="container">
         <div class="row">
-            <div class="span3 column_calc">
```
Comment (bacongobbler) [inline themes/deis/base.html]: just checking that this is intentional, right?

Comment (slack) [inline themes/deis/base.html]: Yep! We will get a fresh paint of coat soon, but wanted to increase the readability now, this gives the content much more room:

2016-05-18 at 09 06

File: src/managing-workflow/production-deployments.md (hunk 19 above)
Comment (slack) [inline src/managing-workflow/production-deployments.md]: reverted

**Comment by bacongobbler** [comment]
I'm happy to just LGTM this for now and start hacking away on any nitpicks I have as follow-ups. Excellent work on the quickstart docs!

**Comment by mboersma** [comment]
> I'm happy to just LGTM this for now and start hacking away on any nitpicks I have as follow-ups. 

Agreed--this is a huge improvement. I'm happy to merge it and I can take a pass at reducing the first-person pronouns in a subsequent PR.

**Comment by bacongobbler** [comment]
needs a rebase now that https://github.com/deis/workflow/pull/239 was merged. Sorry about that!
